### PR TITLE
Fix paused sync file move issue #5949

### PR DESCRIPTION
**Pull request by mrow4a**
@SamuAlfageme can you verify that this fixes issue for ChunkingNG for you? I checked on my local machine and this fixed it.

@guruz I will also add similar thing, but for V1, however there I will check if the file bytes been sent or not. 

Issue #5949

### CODE REVIEW

File: src/libsync/owncloudpropagator.h [hunk 1]
```diff
@@ -214,6 +214,16 @@ class PropagatorCompositeJob : public PropagatorJob
...
             j->abort();
+
+        // Give running jobs max 5 seconds to finish on their own, otherwise
+        // continue (that might result in conflicts owncloud/client/issues/5949)
+        int retriesNo = 5;
+        for(int i = 0; i < retriesNo ; i++) {
+            if (_runningJobs.size() == 0) {
+                break;
+            }
+            Utility::sleep(1);
```
Comment (mrow4a) [inline src/libsync/owncloudpropagator.h]: This is required, because if composite job gets aborted, but one of the job isnt, it will not finalize adding file to journal

File: src/libsync/propagateuploadng.cpp [hunk 2]
```diff
@@ -502,4 +502,16 @@ void PropagateUploadFileNG::slotUploadProgress(qint64 sent, qint64 total)
...
     propagator()->reportProgress(*_item, _sent + sent - total);
 }
+
+
+void PropagateUploadFileNG::abort()
+{
+    foreach (AbstractNetworkJob *job, _jobs) {
+        // Abort only PUT and MKDIR jobs, since abording MoveJob
+        // might result in conflict.
+        if (job->reply() && !job->inherits("OCC::MoveJob")) {
```
Comment (mrow4a) [inline src/libsync/propagateuploadng.cpp]: if it is not MoveJob, just exit. (Wont it result in tmp new chunking directory be left without deletion, is server deleting periodicaly these tmp directories? @DeepDiver1975 )

File: src/libsync/owncloudpropagator.h (hunk 1 above)
Comment (ogoffart) [inline src/libsync/owncloudpropagator.h]: but that's not acceptable. You can't let the main thread sleep like this.
You need to find a better way to wait that the job is actually finished (by listening to the finished signal or so)

File: src/libsync/propagateuploadng.cpp (hunk 2 above)
Comment (ogoffart) [inline src/libsync/propagateuploadng.cpp]: nitpick: use qobject_cast   (more efficient and also compile-time checked)

Also, we shold probzbly decrease the timeout.  For example, if the abort happens because the connection  was detected to be dropped, we should not wait a full HTTP timeout again, IMHO.

Comment (ogoffart) [inline src/libsync/propagateuploadng.cpp]: And to answer your question about stale directory, the next sync will resume the upload, so it won't be stale.

**Comment by guruz** [comment]
> Also the chunkingv1 need to be changed not to abort the last job. This is also what is used when the file is too small to be split in chunks.

I think it's better to hook into how UploadDevice is used and from there control an abort. Then you for sure know you didn't send all bytes and can safely abort.

File: src/libsync/owncloudpropagator.h (hunk 1 above)
Comment (mrow4a) [inline src/libsync/owncloudpropagator.h]: Ohh, true, just tested and it stops the main thread, yep will do that with signals

File: src/libsync/propagateuploadng.cpp (hunk 2 above)
Comment (mrow4a) [inline src/libsync/propagateuploadng.cpp]: Changed to qobject_cast.

**Comment by mrow4a** [comment]
@ogoffart can you have a look on this approach? One disadvantage is that now after paused, MOVE job continues (1st stage):
selection_162

after it is finished, view switches to (2nd stage):
selection_163

One thing is that the button resume sync is still available also in "1st stage", not sure what happens if it still processes MOVE, but someone clicks resume in that time. 

@SamuAlfageme I did manual testing and for ChunkingNG problem is solved.

File: src/libsync/owncloudpropagator.h [hunk 3]
```diff
@@ -227,8 +240,20 @@ private slots:
         return next->scheduleSelfOrChild();
     }
 
+    void terminateJobs() {
+        // Send abort to the subjobs and emit abort finished signal
+        abortJobs();
```
Comment (ogoffart) [inline src/libsync/owncloudpropagator.h]: That's not needed, abort was already sent

So you don't need to put abortJobs in a different function

File: src/libsync/owncloudpropagator.cpp [hunk 4]
```diff
@@ -939,6 +940,11 @@ void PropagateDirectory::slotSubJobsFinished(SyncFileItem::Status status)
     emit finished(status);
 }
 
+void PropagateDirectory::slotAbortFinished()
+{
+    emit abortFinished();
```
Comment (ogoffart) [inline src/libsync/owncloudpropagator.cpp]: You need to count the jobs and only emit the signal when ALL subjobs have finished their abort.

File: src/libsync/owncloudpropagator.h (hunk 3 above)
Comment (mrow4a) [inline src/libsync/owncloudpropagator.h]: The idea here was to force the abort. Thus if after 5 seconds they did not finish, you kill the job just by sending abort jobs again, please have a look on chunkingng implementation

File: src/libsync/owncloudpropagator.cpp (hunk 4 above)
Comment (ogoffart) [inline src/libsync/owncloudpropagator.cpp]: Ah no.  i got confused with CompositeJob.

Well, you don't need a slot for this single line, just connect to SIGNAL(abortFinished())

Comment (mrow4a) [inline src/libsync/owncloudpropagator.cpp]: abortFinished signal is being send after 5 seconds timeout regardless if jobs terminated in that time or not. I give jobs max 5 seconds to finish what they started.

Comment (mrow4a) [inline src/libsync/owncloudpropagator.cpp]: maybe I should change namings, since I see your confusion.

Comment (mrow4a) [inline src/libsync/owncloudpropagator.cpp]: @ogoffart so I can connect signal to signal?

**Comment by ogoffart** [comment]
Looks mostly good.  But don't forget the put in the V1,  this one is even more important as it happens more often.

**Comment by mrow4a** [comment]
Yep, I will do now V1, just wanted to verify that aborting implementation will work for NG (if it will work there, it will probably work everywhere just by implementing abort() in the job)

File: src/libsync/owncloudpropagator.h (hunk 3 above)
Comment (mrow4a) [inline src/libsync/owncloudpropagator.h]: Hmm, maybe in fact I dont need it also... it will roll back to what was before this PR for the jobs which did not terminate within 5sec

**Comment by mrow4a** [comment]
Guys, I tried also the version where CompositeJob waits for aborts from its child, but for some reason with "not aborted MOVE", it was not working properly (looks like a mix of finished and abortFinished signals or even their lack messes things). It took me quite long to debug everything. 

Should I try once again Timeout in the job itself, or keep with version of timeout in  composite job (which will finish sync run no matter what since it does not care if subjob finished abort or not) ?

**Comment by mrow4a** [comment]
@ogoffart @guruz @SamuAlfageme PR ready, fixes both V1 and NG

File: src/libsync/owncloudpropagator.cpp [hunk 5]
```diff
@@ -734,6 +740,20 @@ PropagatorJob::JobParallelism PropagatorCompositeJob::parallelism()
     return FullParallelism;
 }
 
+void PropagatorCompositeJob::abort()
+{
+    foreach (PropagatorJob *j, _runningJobs)
+        j->abort();
```
Comment (guruz) [inline src/libsync/owncloudpropagator.cpp]: please use { and }

File: src/libsync/owncloudpropagator.h [hunk 6]
```diff
@@ -210,11 +210,7 @@ class PropagatorCompositeJob : public PropagatorJob
 
     virtual bool scheduleSelfOrChild() Q_DECL_OVERRIDE;
     virtual JobParallelism parallelism() Q_DECL_OVERRIDE;
-    virtual void abort() Q_DECL_OVERRIDE
-    {
-        foreach (PropagatorJob *j, _runningJobs)
-            j->abort();
-    }
+    virtual void abort() Q_DECL_OVERRIDE;
```
Comment (guruz) [inline src/libsync/owncloudpropagator.h]: Please also document here that the abort() might not be synchronous/immediate

**Comment by mrow4a** [comment]
> I'm missing a bit of info what happens in a situation where an immediate abort does not work. What will > happen after those 5000msec? It will go to emitFinished but what happens with the running stuff?

Good question, in my observation when job finishes the sync run is already gone, thus whatever finishes after that time is ignored.

I think I will try to implement "forced abort".

**Comment by guruz** [comment]
I'm not sure, maybe the QNetworkReplys are parented to the Jobs which get deleted when SyncEngine is deleted.
(Which is the force abort, basically)

**Comment by mrow4a** [comment]
@guruz how to check it? @jturcotte

**Comment by jturcotte** [comment]
Maybe just call `abort` in the code where you think the issue could appear.

For example in tests we have `FakeFolder::execUntilItemCompleted` that allows you to something right after an `itemCompleted` signal. But you could also just test this manually in the code with a QTimer or something like that.

File: src/libsync/owncloudpropagator.h (hunk 6 above)
Comment (ogoffart) [inline src/libsync/owncloudpropagator.h]: Actually, the PropagatorJob::abort documentation should state that this function must emit abortFinished

File: src/libsync/owncloudpropagator.h [hunk 7]
```diff
@@ -261,6 +261,9 @@ class OWNCLOUDSYNC_EXPORT PropagateDirectory : public PropagatorJob
     virtual JobParallelism parallelism() Q_DECL_OVERRIDE;
     virtual void abort() Q_DECL_OVERRIDE
     {
+        // Abort first job and sub jobs
+        // Finished abort will be announced via abortFinished()
+        // (look contructor implementation)
```
Comment (ogoffart) [inline src/libsync/owncloudpropagator.h]: (Maybe you can do the connection here.)

Here as well you need to have a counter to make sure that both jobs sent their abortFinished.

File: src/libsync/propagateuploadng.cpp [hunk 8]
```diff
@@ -502,4 +502,20 @@ void PropagateUploadFileNG::slotUploadProgress(qint64 sent, qint64 total)
...
 }
+
+
+void PropagateUploadFileNG::abort()
+{
+    foreach (AbstractNetworkJob *job, _jobs) {
+        // Dont abort MoveJob since it might result in conflict,
+        // only PUT and MKDIR jobs can be safely aborted
+        if (qobject_cast<MoveJob *>(job)){
+            continue;
```
Comment (ogoffart) [inline src/libsync/propagateuploadng.cpp]: I miss the point where abortFinished is sent when that jobs finished

File: src/libsync/propagateuploadv1.cpp [hunk 9]
```diff
@@ -360,4 +358,21 @@ void PropagateUploadFileV1::slotUploadProgress(qint64 sent, qint64 total)
...
 }
+
+void PropagateUploadFileV1::abort()
+{
+    foreach (AbstractNetworkJob *job, _jobs) {
+        // Dont abort final PUT which uploaded its data,
+        // since this might result in conflicts
+        if (PUTFileJob *putJob = qobject_cast<PUTFileJob *>(job)){
+            if (isFinalChunk() && putJob->device()->atEnd()) {
+                continue;
```
Comment (ogoffart) [inline src/libsync/propagateuploadv1.cpp]: Same thing, We need to emit abortFinished when the job finishes.

**Comment by mrow4a** [comment]
@ogoffart I did that, and for some reason it was not working properly e.g. finished signal never went back after aborting, and other issues... I can give a try once again on that approach.. 

I just discovered that I also have to be careful not to break abort mechanism on network error with move, since it will finish in infinite aborts loop (it will retry all the time).

File: src/libsync/owncloudpropagator.h [hunk 10]
```diff
@@ -394,11 +399,13 @@ class OwncloudPropagator : public QObject
     {
         _abortRequested.fetchAndStoreOrdered(true);
         if (_rootJob) {
-            // We're possibly already in an item's finished stack
```
Comment (ogoffart) [inline src/libsync/owncloudpropagator.h]: keep this comment, it explain why we do a QueuedConnection

File: src/libsync/owncloudpropagator.h [hunk 11]
```diff
@@ -394,11 +399,13 @@ class OwncloudPropagator : public QObject
...
+            // Invoke abort and wait for abort finish to finalize
+            connect(_rootJob.data(), SIGNAL(abortFinished()), this, SLOT(slotAbortFinished()));
```
Comment (ogoffart) [inline src/libsync/owncloudpropagator.h]: maybe here we could add a timeout in case the jobs takes too long to finish .

File: src/libsync/owncloudpropagator.h [hunk 12]
```diff
@@ -395,10 +442,15 @@ class OwncloudPropagator : public QObject
         _abortRequested.fetchAndStoreOrdered(true);
         if (_rootJob) {
             // We're possibly already in an item's finished stack
-            QMetaObject::invokeMethod(_rootJob.data(), "abort", Qt::QueuedConnection);
+            // Invoke abort and wait for abort finish
+            // to finalize asynchronously with abortFinished
+            connect(_rootJob.data(), SIGNAL(abortFinished()), this, SLOT(slotAbortFinished()));
+            QTimer::singleShot(jobTimeoutDuration(), this, SLOT(abortTimeout()));
```
Comment (mrow4a) [inline src/libsync/owncloudpropagator.h]: I give the jobs 5sec to finish on their own asychronously, otherwise I force them to abort synchronously and invoke emitFinished

**Comment by mrow4a** [comment]
@ogoffart @guruz ready for review

**Comment by mrow4a** [comment]
Any input here @ogoffart @guruz @jturcotte ? Should I split it to smaller PRs?

File: src/libsync/owncloudpropagator.cpp [hunk 13]
```diff
@@ -410,6 +410,12 @@ quint64 OwncloudPropagator::smallFileSize()
     return smallFileSize;
 }
 
+quint64 OwncloudPropagator::jobTimeoutDuration()
```
Comment (ogoffart) [inline src/libsync/owncloudpropagator.cpp]: You don't need this function. And that's not even a good name. (you don't mention it is for abort, you don't mention what unit it is)

Since it is only used once, just have the constant in OwncloudPropagator::abort

(Note: in new code, the time constant should probably be in Ideally, in new code, the type of such constant should be `std::chrono::milliseconds` so the unit is in the type )

File: src/libsync/owncloudpropagator.h [hunk 14]
```diff
@@ -419,6 +461,17 @@ class OwncloudPropagator : public QObject
 
 private slots:
 
+    void abortTimeout()
+    {
+        // Abort synchronously (asyncAbort set to false) and finish
+        QMetaObject::invokeMethod(_rootJob.data(), "abort", Qt::QueuedConnection, Q_ARG(bool, false));
```
Comment (ogoffart) [inline src/libsync/owncloudpropagator.h]: Is there a reason you use a QueuedConnection here?
the abort(false) will be called after the slotAbortFinished.
If it is required, add a comment why we don't just call abort(false),   if it is not required, just make a call to abort(false)



(also that's a short function, it might be cleaner to use a lambda for it)

File: src/libsync/owncloudpropagator.h [hunk 15]
```diff
@@ -419,6 +461,17 @@ class OwncloudPropagator : public QObject
...
+        slotAbortFinished();
+    }
+
+    void slotAbortFinished() {
+        QMetaObject::invokeMethod(this, "emitFinished", Qt::QueuedConnection, Q_ARG(SyncFileItem::Status, SyncFileItem::NormalError));
```
Comment (ogoffart) [inline src/libsync/owncloudpropagator.h]: please add a comment on why we need a queued connection

**Comment by ogoffart** [comment]
Also please rebase in a single commit.

**Comment by mrow4a** [comment]
Ok, did what requested @ogoffart :
- [x] Used QT5 connect style
- [x] Updated/Deleted Queued Connection where required
- [x] Test New Chunking MOVE abort 
- [x] Test New Chunking MOVE abort with timeout of 5s exceeded -> it starts downloading file after forced abort as in the synchronous scenario
- [x] Test OLD Chunking FINAL PUT abort 
- [x] Test New Chunking FINAL PUT abort with timeout of 5s exceeded -> it starts downloading file after forced abort as in the synchronous scenario

**Comment by mrow4a** [comment]
Why this PR started failing continuous-integration test for ChunkingNG? I dont get the error and it makes no sense

**Comment by jturcotte** [comment]
@mrow4a A few lines lower:
```
16: 09-05 14:45:07:141 [ info sync.engine ]:	There are 487604224 bytes available at "/tmp/ChunkingNgTest-ULWWGt/" and at least 50000000 are required
```
@jnweiger said there is only 1GB in /tmp on that machine. Maybe we could lower that threshold for auto tests so that we don't have to do manual cleanup that often.

File: src/libsync/owncloudpropagator.h [hunk 16]
```diff
@@ -98,7 +98,14 @@ class PropagatorJob : public QObject
     virtual qint64 committedDiskSpace() const { return 0; }
 
 public slots:
-    virtual void abort() {}
+    /*
+     * Asynchronous abort requires emit of abortFinished() signal,
+     * while synchronous is expected to abort immedietaly.
+    */
+    virtual void abort(const bool asyncAbort) {
```
Comment (ogoffart) [inline src/libsync/owncloudpropagator.h]: parameters passed by value should not be const.   It should simply be `bool asyncAbort`


(Ideally, even, this should not be a bool, but an enum { Synchronously, Asynchronously } )

Comment (mrow4a) [reply src/libsync/owncloudpropagator.h]: omg, I overlooked it, of course it makes no sense, thanks! about enum, I am not sure, but definitely that would improve readability.

File: src/libsync/owncloudpropagator.cpp [hunk 17]
```diff
@@ -891,7 +901,8 @@ void PropagateDirectory::slotFirstJobFinished(SyncFileItem::Status status)
 
     if (status != SyncFileItem::Success && status != SyncFileItem::Restoration) {
         if (_state != Finished) {
-            abort();
+            // Synchronously abort
+            abort(AbortType::Synchronous);
```
Comment (mrow4a) [inline src/libsync/owncloudpropagator.cpp]: Now, instead of boolean, there is enum

File: src/libsync/owncloudpropagator.h [hunk 18]
```diff
@@ -98,7 +103,14 @@ class PropagatorJob : public QObject
     virtual qint64 committedDiskSpace() const { return 0; }
 
 public slots:
-    virtual void abort() {}
+    /*
+     * Asynchronous abort requires emit of abortFinished() signal,
+     * while synchronous is expected to abort immedietaly.
+    */
+    virtual void abort(PropagatorJob::AbortType abortType) {
```
Comment (mrow4a) [inline src/libsync/owncloudpropagator.h]: function requires "full definition" due to the fact that we call it using invokeMethod. Without it it thrown ```QMetaObject::invokeMethod: No such method OCC::PropagateDirectory::abort(PropagatorJob::AbortType)
Candidates are:
    abort(AbortType)```

File: src/libsync/owncloudpropagator.h [hunk 19]
```diff
@@ -324,6 +364,7 @@ class OwncloudPropagator : public QObject
         , _chunkSize(10 * 1000 * 1000) // 10 MB, overridden in setSyncOptions
         , _account(account)
     {
+        qRegisterMetaType<PropagatorJob::AbortType>("PropagatorJob::AbortType");
```
Comment (mrow4a) [inline src/libsync/owncloudpropagator.h]: It needs to register metatype, othewise ```09-06 16:41:33:780 [ info sync.engine ]:	Aborting sync
09-06 16:41:33:780 [ warning default ]:	QMetaMethod::invoke: Unable to handle unregistered datatype 'PropagatorJob::AbortType'
09-06 16:41:33:780 [ fatal default ]:	ASSERT: "invoked" in file /home/mrow4a/Projects/client/src/libsync/owncloudpropagator.h, line 444```

**Comment by ckamm** [comment]
@jturcotte About
```
16: 09-05 14:45:07:141 [ info sync.engine ]:	There are 487604224 bytes available at "/tmp/ChunkingNgTest-ULWWGt/" and at least 50000000 are required
```
Note that this says: 487MB are available and 50MB are required. I've also stumbled over this and will adjust the message to make it more clear whether it's informative or an error.

**Comment by ckamm** [comment]
I've rebased to master and will review the current state to help move this forward.

**Comment by ckamm** [comment]
Looks good. My remaining worry is about what happens when abort(Async) is called twice. Do we guard against this somewhere?

**Comment by mrow4a** [comment]
I mean, if you call it again, it will basically let unfinished job run (as with previous call of abort(Async)), the other jobs are synchronously aborted. The only difference is, that now you might have 2 timers (timeout timers) running in parallel, and you add another connects on signals. @ogoffart is that a problem?

**Comment by ckamm** [comment]
@mrow4a I added a minor guard against it that seems sufficient for me.

**Comment by ckamm** [comment]
@mrow4a @guruz @ogoffart Are there any remaining concerns (also about the code I added)? Can we merge this?

File: src/libsync/propagateupload.h [hunk 20]
```diff
@@ -320,6 +320,10 @@ class PropagateUploadFileV1 : public PropagateUploadFileCommon
 public:
     PropagateUploadFileV1(OwncloudPropagator *propagator, const SyncFileItemPtr &item)
         : PropagateUploadFileCommon(propagator, item)
+        , _startChunk(0)
+        , _currentChunk(0)
```
Comment (ogoffart) [inline src/libsync/propagateupload.h]: Note: this could be initialized in the header file.

File: src/libsync/propagateuploadv1.cpp [hunk 21]
```diff
@@ -363,8 +363,10 @@ void PropagateUploadFileV1::abort(PropagatorJob::AbortType abortType)
             // dont abort final PUT which uploaded its data,
             // since this might result in conflicts
             if (PUTFileJob *putJob = qobject_cast<PUTFileJob *>(job)){
-                if (abortType == AbortType::Asynchronous && (((_currentChunk + _startChunk) % _chunkCount) == 0)
-                        && putJob->device()->atEnd()) {
+                if (abortType == AbortType::Asynchronous
+                    && _chunkCount > 0
```
Comment (ogoffart) [inline src/libsync/propagateuploadv1.cpp]: I wonder how chunk count can be 0

File: src/libsync/propagateuploadv1.cpp [hunk 22]
```diff
@@ -350,4 +350,31 @@ void PropagateUploadFileV1::slotUploadProgress(qint64 sent, qint64 total)
...
+
+    // Abort all jobs (if there are any left), except final PUT
+    foreach (AbstractNetworkJob *job, _jobs) {
+        if (job->reply()) {
+            // If asynchronous abort allowed,
+            // dont abort final PUT which uploaded its data,
+            // since this might result in conflicts
+            if (PUTFileJob *putJob = qobject_cast<PUTFileJob *>(job)){
+                if (abortType == AbortType::Asynchronous
+                    && _chunkCount > 0
```
Comment (mrow4a) [inline src/libsync/propagateuploadv1.cpp]: Can you elaborate more here?

**Comment by mrow4a** [comment]
Any progress here?

File: src/libsync/propagateupload.h (hunk 20 above)
Comment (mrow4a) [reply src/libsync/propagateupload.h]: Should I step in and fix it, or @ckamm ?

File: src/libsync/propagateuploadv1.cpp (hunk 22 above)
Comment (mrow4a) [reply src/libsync/propagateuploadv1.cpp]: I guess this is safety check right?

File: src/libsync/propagateupload.h (hunk 20 above)
Comment (ckamm) [reply src/libsync/propagateupload.h]: Sorry, I missed the notifications for this. Looking now.

File: src/libsync/propagateuploadv1.cpp (hunk 21 above)
Comment (ckamm) [reply src/libsync/propagateuploadv1.cpp]: `chunkCount` gets set by `doStartUpload` and the job's `abort` can be called before we get there.

File: src/libsync/propagateuploadv1.cpp (hunk 22 above)
Comment (ckamm) [reply src/libsync/propagateuploadv1.cpp]: See above. I was getting crashes because of zero chunkCount when I aborted jobs quickly.

File: src/libsync/propagateupload.h (hunk 20 above)
Comment (ckamm) [reply src/libsync/propagateupload.h]: @mrow4a (in general, feel free to just fix issues if I'm unresponsive!)

**Comment by ckamm** [comment]
@mrow4a I think you can merge this now

File: src/libsync/propagateuploadv1.cpp (hunk 22 above)
Comment (mrow4a) [reply src/libsync/propagateuploadv1.cpp]: Ohh that would make sense did not try indeed this corner case.

**Comment by mrow4a** [comment]
What is happening with Jenkins build? Did you run tests locally?

**Comment by ckamm** [comment]
No, I hadn't. The test that fails to compile on jenkins works locally. But I'll check what's going on there.

**Comment by ckamm** [comment]
Jenkins should pass now, let's wait and see.

**Comment by ckamm** [comment]
@DeepDiver1975 I think we might want this in 2.4, but jenkins failed in something after running the tests. Is this a transient issue where retriggering jenkins would help? Could you do that?

**Comment by ogoffart** [comment]
We anyway need to merge this to the 2.4 branch.

**Comment by ckamm** [comment]
Targeted to 2.4 and merged!
